src/csharp/Microsoft.Spark/Sql/CastUnpickledItems.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Spark.Sql.Types;
+
+namespace Microsoft.Spark.Sql
+{
+    /// <summary>
+    /// Cast UnpickledItems.
+    /// </summary>
+    internal static class CastUnpickledItems
+    {
+        /// <summary>
+        /// Cast unpickledItems from arraylist to the appropriate array.
+        /// </summary>
+        /// <param name="unpickledItems">Unpickled objects to be cast as necessary.</param>
+        /// <returns>Unpickled objects after casting.</returns>
+        public static object[] Cast(object unpickledItems)
+        {
+            var castUnpickledItems = new List<object>();
+            foreach (object obj in (object[])unpickledItems)
+            {
+                castUnpickledItems.Add(
+                    (obj.GetType() == typeof(RowConstructor)) ? obj : CastArray(obj));
```

Comment:
I am not sure if falling back to `CastArray` always works.

Reply:
I have updated the function name accordingly. Sorry for the confusion. If not `RowConstructor`, it will fall into `object []`. `CastHelper` will help decide and cast `object[]` as needed.